### PR TITLE
Laboratory work #4, Ruslana Zobnina - 22FPL2.

### PR DESCRIPTION
**Pull request by RuslanaZobnina**

### CODE REVIEW

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_generate_by_ngrams/target_score.txt [hunk 1]
```diff
@@ -1 +1 @@
-10
+6
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/target_score.txt:1]: revert

File: lab_4_fill_words_by_ngrams/main.py [hunk 2]
```diff
@@ -28,6 +30,19 @@ def _tokenize(self, text: str) -> tuple[str, ...]:  # type: ignore
...
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
+        if not isinstance(text, str) or not text:
+            raise ValueError('Type input is inappropriate or input argument is empty.')
+
+        tokens = ''
+        end_of_word_punctuation = '?.!'
+        for word in text.lower():
+            if word in end_of_word_punctuation:
+                tokens += ' ' + self._end_of_word_token
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:40]: fstring

File: lab_4_fill_words_by_ngrams/main.py [hunk 3]
```diff
@@ -98,6 +132,31 @@ def run(self, seq_len: int, prompt: str) -> str:  # type: ignore
...
+        if not encoded:
+            raise ValueError("it wasn't possible to encode the prompt")
+
+        while seq_len > 0:
+            tokens = self._model.generate_next_token(encoded)
+            if not tokens:
+                raise ValueError('this method returns None')
+
+            break
+
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:152]: too

File: lab_4_fill_words_by_ngrams/main.py [hunk 4]
```diff
@@ -98,6 +132,31 @@ def run(self, seq_len: int, prompt: str) -> str:  # type: ignore
...
+
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:153]: oo

File: lab_4_fill_words_by_ngrams/main.py [hunk 5]
```diff
@@ -98,6 +132,31 @@ def run(self, seq_len: int, prompt: str) -> str:  # type: ignore
...
+
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:154]: oo

File: lab_4_fill_words_by_ngrams/main.py [hunk 6]
```diff
@@ -98,6 +132,31 @@ def run(self, seq_len: int, prompt: str) -> str:  # type: ignore
...
+
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:155]: o_O

File: lab_4_fill_words_by_ngrams/main.py [hunk 7]
```diff
@@ -98,6 +132,31 @@ def run(self, seq_len: int, prompt: str) -> str:  # type: ignore
...
+
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:156]: O_o

File: lab_4_fill_words_by_ngrams/main.py [hunk 8]
```diff
@@ -98,6 +132,31 @@ def run(self, seq_len: int, prompt: str) -> str:  # type: ignore
...
+
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:157]: O_O

File: lab_4_fill_words_by_ngrams/main.py [hunk 9]
```diff
@@ -98,6 +132,31 @@ def run(self, seq_len: int, prompt: str) -> str:  # type: ignore
...
+
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:158]: many empty lines

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
What about start.py?

**Review by artyomtugaryov — COMMENTED** [review]
Very good

**Comment by vitazaytseva** [comment]
Well done! Good job!

**Pull request closed without merging**